source/slang/slang-capability.h

```diff
@@ -189,62 +131,172 @@ struct CapabilitySet
+
+    inline void addToTargetCapabilityWithTargetAndStageAtom(const CapabilityName& target, const CapabilityName& stage, const ArrayView<CapabilityName>& canonicalRepresentation);
+    inline void addToTargetCapabilityWithTargetAndOrStageAtom(const CapabilityName& target, const CapabilityName& stage, const ArrayView<CapabilityName>& canonicalRepresentation);
+    inline void addToTargetCapabilityWithStageAtom(const CapabilityName& stage, const ArrayView<CapabilityName>& canonicalRepresentation);
+    inline void addToTargetCapabilitesWithCanonicalRepresentation(const ArrayView<CapabilityName>& atom);
+    inline void addUnexpandedCapabilites(const CapabilityName& atom);
+    
+    CapabilityTargetSets& getCapabilityTargetSets() { return m_targetSets; }
+    const CapabilityTargetSets& getCapabilityTargetSets() const { return m_targetSets; }
+
+    struct AtomSets
+    {
+        struct Iterator
+        {
+        private:
+            const CapabilityTargetSets* context;
+            decltype(context->begin()) targetNode;
+            decltype((*targetNode).second.shaderStageSets.begin()) stageNode;
+            decltype((*stageNode).second.disjointSets.begin()) disjointSetNode;
+
+        public:
+            bool isValid() const
+            {
+                return disjointSetNode.current;
+            }
+            const CapabilityAtomSet& operator*() const
+            {
+                return (*this->disjointSetNode);
+            }
+            const CapabilityAtomSet* operator->() const
+            {
+                return &(*this->disjointSetNode);
+            }
+            bool operator==(const Iterator& other) const
+            {
+                return other.context == this->context
+                    && other.targetNode == this->targetNode
+                    ;
+            }
+            bool operator!=(const Iterator& other) const
+            {
+                return !(other == *this);
+            }
+
+            Iterator& operator++()
+            {
+                this->disjointSetNode++;
+                if (this->disjointSetNode == (*this->stageNode).second.disjointSets.end())
+                {
+nullDisjointSetNode:
+                    this->stageNode++;
+                    if (this->stageNode == (*this->targetNode).second.shaderStageSets.end())
+                    {
+nullStageSetNode:
+                        this->targetNode++;
+                        if (this->targetNode == this->context->end())
+                        {
+                            this->stageNode = {};
+                            this->disjointSetNode = {};
+                            return *this;
+                        }
+                        this->stageNode = (*this->targetNode).second.shaderStageSets.begin();
+                        if (this->stageNode == (*this->targetNode).second.shaderStageSets.end())
+                            goto nullStageSetNode;
```

Comment:
`goto`s like this can be rewritten as a `for(;;)` at the label and a `continue` at the jump.